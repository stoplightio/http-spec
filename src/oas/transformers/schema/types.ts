import type { JSONSchema7 } from 'json-schema';
import type { SchemaObject } from 'openapi3-ts/src/model/OpenApi';
import type { Schema } from 'swagger-schema-official';

export type OASSchemaObject = SchemaObject | Schema;

export type Converter = (schema: OASSchemaObject | JSONSchema7) => void;
