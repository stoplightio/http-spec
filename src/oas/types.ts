import { DeepPartial, IHttpOperation, IHttpService } from '@stoplight/types';
import { OpenAPIObject } from 'openapi3-ts';
import { Spec } from 'swagger-schema-official';

export const enum OasVersion {
  OAS2 = 2,
  OAS3 = 3,
}

/**
 * Service
 */

export interface ITransformServiceOpts<T> {
  document: DeepPartial<T>;
}

export type HttpServiceTransformer<T> = (opts: T) => IHttpService;

export type Oas2TransformServiceOpts = ITransformServiceOpts<Spec>;
export type Oas3TransformServiceOpts = ITransformServiceOpts<OpenAPIObject>;
export type Oas2HttpServiceTransformer = HttpServiceTransformer<Oas2TransformServiceOpts>;
export type Oas3HttpServiceTransformer = HttpServiceTransformer<Oas3TransformServiceOpts>;

/**
 * Operation
 */

export interface ITransformOperationOpts<T> {
  document: DeepPartial<T>;
  path: string;
  method: string;
}

export type HttpOperationTransformer<T> = (opts: T) => IHttpOperation;

export type Oas2TransformOperationOpts = ITransformOperationOpts<Spec>;
export type Oas3TransformOperationOpts = ITransformOperationOpts<OpenAPIObject>;
export type Oas2HttpOperationTransformer = HttpOperationTransformer<Oas2TransformOperationOpts>;
export type Oas3HttpOperationTransformer = HttpOperationTransformer<Oas3TransformOperationOpts>;
