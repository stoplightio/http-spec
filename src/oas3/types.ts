import { DeepPartial } from '@stoplight/types';
import { OpenAPIObject as _OpenAPIObject } from 'openapi3-ts';
import { PathItemObject } from 'openapi3-ts/src/model/OpenApi';
import { ISpecificationExtension } from 'openapi3-ts/src/model/SpecificationExtension';

import { TranslateFunction } from '../types';

export type Oas3TranslateFunction<P extends unknown[], R extends unknown = unknown> = TranslateFunction<
  DeepPartial<OpenAPIObject>,
  P,
  R
>;

export interface OpenAPIObject extends _OpenAPIObject {
  webhooks?: WebhooksObject;
}

export interface WebhooksObject extends ISpecificationExtension {
  [name: string]: PathItemObject;
}
