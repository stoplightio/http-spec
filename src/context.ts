import { hasRef, isLocalRef } from '@stoplight/json';
// @ts-ignore
import * as fnv from 'fnv-plus';

import type {
  AvailableContext,
  Fragment,
  IdGenerator,
  RefResolver,
  TransformerContext,
  TranslateFunction,
} from './types';

export const DEFAULT_ID_GENERATOR: IdGenerator = input => fnv.fast1a32hex(input);

export function createContext<T extends Record<string, unknown>>(
  document: T,
  resolveRef: RefResolver<T>,
  generateId: IdGenerator,
): TransformerContext<T> {
  let context: AvailableContext = 'service';
  return {
    document,
    get context() {
      return context;
    },
    set context(value) {
      context = value;
      if (value !== 'operation') {
        this.parentId = this.ids[value];
      }
    },
    maybeResolveLocalRef(target: unknown) {
      if (hasRef(target) && isLocalRef(target.$ref)) {
        try {
          return resolveRef.call(this, target);
        } catch {
          return target;
        }
      }

      return target;
    },
    generateId(value) {
      this.parentId = generateId(value);
      return this.parentId;
    },
    ids: {
      service: '',
      path: '',
      operation: '',
    },
    parentId: '',
  };
}

export function withContext<F extends TranslateFunction<any, any[]> = TranslateFunction<Fragment, unknown[]>>(
  fn: F,
): F {
  return <F>function (...args: Parameters<F>) {
    const { context, parentId } = this;
    const result = fn.apply(this, args);
    this.context = context;
    this.parentId = parentId;
    return result;
  };
}
