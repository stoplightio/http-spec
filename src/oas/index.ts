export * from './operation';
export * from './types';
