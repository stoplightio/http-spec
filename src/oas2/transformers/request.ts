import { DeepPartial, IHttpOperationRequest } from '@stoplight/types';
import { BodyParameter, FormDataParameter, Parameter, Spec } from 'swagger-schema-official';

import {
  translateFromFormDataParameters,
  translateToBodyParameter,
  translateToHeaderParam,
  translateToPathParameter,
  translateToQueryParameter,
} from './params';

export function translateToRequest(
  document: DeepPartial<Spec>,
  parameters: Parameter[],
  consumes: string[],
): IHttpOperationRequest {
  const bodyParameters = parameters.filter((p): p is BodyParameter => p.in === 'body');
  const formDataParameters = parameters.filter((p): p is FormDataParameter => p.in === 'formData');
  const request: IHttpOperationRequest = {};

  // if 'body' and 'form data' defined prefer 'body'
  if (!!bodyParameters.length) {
    // There can be only one body parameter (taking first one)
    request.body = translateToBodyParameter(document, bodyParameters[0], consumes);
  } else if (!!formDataParameters.length) {
    request.body = translateFromFormDataParameters(document, formDataParameters, consumes);
  }

  return parameters.reduce(createReduceRemainingParameters(document), request);
}

function createReduceRemainingParameters(document: DeepPartial<Spec>) {
  return function (request: IHttpOperationRequest, parameter: Parameter) {
    if (parameter.in === 'query') {
      const queryParameter = translateToQueryParameter(document, parameter);
      request.query = (request.query || []).concat(queryParameter);
    } else if (parameter.in === 'path') {
      const pathParameter = translateToPathParameter(document, parameter);
      request.path = (request.path || []).concat(pathParameter);
    } else if (parameter.in === 'header') {
      const headerParameter = translateToHeaderParam(document, parameter);
      request.headers = (request.headers || []).concat(headerParameter);
    }
    return request;
  };
}
