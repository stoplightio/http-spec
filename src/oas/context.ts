// @ts-expect-error: no types
import * as fnv from 'fnv-plus';

import { createContext as _createContext } from '../context';
import type { Fragment, IdGenerator } from '../types';
import { resolveRef } from './resolver';

const HASH_32_KINDS = ['http_operation', 'schema'];

function getKind(input: string) {
  const i = input.indexOf('-');
  return input.slice(0, i);
}

const DEFAULT_ID_GENERATOR: IdGenerator = input => {
  const kind = getKind(input);
  return HASH_32_KINDS.some(k => kind === k) ? fnv.fast1a32hex(input) : fnv.fast1a52hex(input);
};

export function createContext<T extends Fragment>(document: T) {
  return _createContext<T>(document, resolveRef, DEFAULT_ID_GENERATOR);
}
