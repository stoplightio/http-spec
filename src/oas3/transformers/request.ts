import { isPlainObject } from '@stoplight/json';
import type {
  IHttpHeaderParam,
  IHttpOperationRequest,
  IHttpOperationRequestBody,
  IHttpParam,
  INodeExample,
  INodeExternalExample,
  Optional,
} from '@stoplight/types';
import { HttpParamStyles } from '@stoplight/types';
import type { JSONSchema7 } from 'json-schema';
import type { ParameterObject } from 'openapi3-ts';
import pickBy = require('lodash.pickby');

import { isNonNullable, isString } from '../../guards';
import { OasVersion } from '../../oas';
import { getValidOasParameters } from '../../oas/accessors';
import { isValidParamStyle } from '../../oas/guards';
import { translateSchemaObject } from '../../oas/transformers/schema';
import type { ArrayCallbackParameters, Fragment } from '../../types';
import { entries } from '../../utils';
import { isRequestBodyObject } from '../guards';
import type { Oas3TranslateFunction } from '../types';
import { translateMediaTypeObject } from './content';

export const translateRequestBody: Oas3TranslateFunction<[requestBodyObject: unknown], IHttpOperationRequestBody> =
  function (requestBodyObject) {
    const resolvedRequestBodyObject = this.maybeResolveLocalRef(requestBodyObject);

    if (isRequestBodyObject(resolvedRequestBodyObject)) {
      return {
        contents: entries(resolvedRequestBodyObject.content).map(translateMediaTypeObject, this).filter(isNonNullable),

        ...pickBy(
          {
            required: resolvedRequestBodyObject.required,
            description: resolvedRequestBodyObject.description,
          },
          isNonNullable,
        ),
      };
    }

    return { contents: [] };
  };

const translateParameterObjectSchema: Oas3TranslateFunction<
  [parameterObject: Fragment],
  Optional<JSONSchema7>
> = function (parameterObject) {
  if (!isPlainObject(parameterObject.schema)) return;

  return translateSchemaObject.call(this, {
    ...parameterObject.schema,
    ...('example' in parameterObject ? { example: parameterObject.example } : null),
  });
};

const translateToExample: Oas3TranslateFunction<
  ArrayCallbackParameters<[key: string, example: unknown]>,
  Optional<INodeExample | INodeExternalExample>
> = function ([key, example]) {
  if (!isPlainObject(example)) return;

  if (!('value' in example) && typeof example.externalValue !== 'string') return;

  return {
    key,

    ...(typeof example.externalValue === 'string'
      ? { externalValue: example.externalValue }
      : { value: example.value }),

    ...pickBy(
      {
        summary: example.summary,
        description: example.description,
      },
      isString,
    ),
  };
};

export const translateParameterObject: Oas3TranslateFunction<[parameterObject: ParameterObject], IHttpParam> =
  function (parameterObject) {
    const examples = entries(parameterObject.examples).map(translateToExample, this).filter(isNonNullable);

    const hasDefaultExample = examples.some(({ key }) => key.includes('default'));
    const schema = translateParameterObjectSchema.call(this, parameterObject);

    return {
      name: parameterObject.name,
      deprecated: !!parameterObject.deprecated,
      style: isValidParamStyle(parameterObject.style) ? parameterObject.style : HttpParamStyles.Simple,
      explode: !!(parameterObject.explode ?? parameterObject.style === HttpParamStyles.Form),
      examples:
        'example' in parameterObject && !hasDefaultExample
          ? [{ key: 'default', value: parameterObject.example }, ...examples]
          : examples,

      ...pickBy(
        {
          description: parameterObject.description,
        },
        isString,
      ),

      ...pickBy(
        {
          schema,
          content: parameterObject.content,
        },
        isPlainObject,
      ),
    };
  };

export const translateToRequest: Oas3TranslateFunction<[path: Fragment, operation: Fragment], IHttpOperationRequest> =
  function (path, operation) {
    const params: Omit<IHttpOperationRequest, 'header'> & { header: IHttpHeaderParam[] } = {
      header: [],
      query: [],
      cookie: [],
      path: [],
    };

    for (const param of getValidOasParameters(this.document, OasVersion.OAS3, operation.parameters, path.parameters)) {
      const { in: key } = param;
      const target = params[key];
      if (!Array.isArray(target)) continue;

      target.push(translateParameterObject.call(this, param) as any);
    }

    return {
      body: translateRequestBody.call(this, operation?.requestBody),
      headers: params.header,
      query: params.query,
      cookie: params.cookie,
      path: params.path,
    };
  };
