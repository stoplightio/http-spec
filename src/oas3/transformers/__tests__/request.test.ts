import { IHttpQueryParam } from '@stoplight/types';

import { createContext } from '../../../oas/context';
import { translateToRequest as _translateToRequest } from '../request';

const translateToRequest = (path: Record<string, unknown>, operation: Record<string, unknown>, operationId: string) =>
  _translateToRequest.call(createContext({}), path, operation, { id: operationId });

describe('translateOas3ToRequest', () => {
  it('given no request body should translate parameters', () => {
    const operation = {
      parameters: [
        {
          name: 'param-name-2',
          in: 'query',
          description: 'descr',
          content: {
            'content-b': {
              schema: {},
            },
          },
        },
        {
          name: 'param-name-3',
          in: 'header',
          description: 'descr',
          content: {
            'content-c': {
              schema: {},
            },
          },
        },
      ],
    };

    const path = {
      parameters: [
        {
          name: 'param-name-1',
          in: 'query',
          description: 'descr',
          deprecated: true,
          content: {
            'content-a': {
              schema: {},
            },
          },
        },
      ],
      get: operation,
    };

    expect(translateToRequest(path, operation, 'get')).toMatchSnapshot({
      headers: [
        {
          id: expect.any(String),
        },
      ],
      query: [
        {
          id: expect.any(String),
        },
        {
          id: expect.any(String),
        },
      ],
    });
  });

  it('give a request body should translate it', () => {
    const operation = {
      requestBody: {
        description: 'descr',
        required: true,
        content: {
          'content-a': {
            schema: {
              deprecated: true,
            },
          },
        },
      },
    };

    const path = {
      post: operation,
    };

    expect(translateToRequest(path, operation, 'post')).toMatchSnapshot({
      body: {
        id: expect.any(String),
        contents: [
          {
            id: expect.any(String),
            schema: {
              'x-stoplight': {
                id: expect.any(String),
              },
            },
          },
        ],
      },
    });
  });

  it('given path-defined parameters should create unique ids', () => {
    const getOperation = {
      parameters: [],
    };
    const postOperation = {
      parameters: [],
    };

    const path = {
      parameters: [
        {
          name: 'param-name-1',
          in: 'query',
          description: 'descr',
          deprecated: true,
          content: {
            'content-a': {
              schema: {},
            },
          },
        },
      ],
      get: getOperation,
      post: postOperation,
    };

    const getRequest = translateToRequest(path, getOperation, 'get');
    const postRequest = translateToRequest(path, postOperation, 'post');
    const getQueryId = (getRequest.query?.[0] as IHttpQueryParam<true>).id;
    const postQueryId = (postRequest.query?.[0] as IHttpQueryParam<true>).id;

    expect(getQueryId).not.toEqual(postQueryId);
  });
});
