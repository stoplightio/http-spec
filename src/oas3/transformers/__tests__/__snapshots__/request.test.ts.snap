// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translateOas3ToRequest give a request body should translate it 1`] = `
Object {
  "body": Object {
    "contents": Array [
      "translateMediaTypeObject({}, {\\"schema\\":{\\"deprecated\\":true}}, content-a, {\\"content-a\\":{\\"schema\\":{\\"deprecated\\":true}}})",
    ],
    "description": "descr",
    "required": true,
  },
  "cookie": Array [],
  "headers": Array [],
  "path": Array [],
  "query": Array [],
}
`;

exports[`translateOas3ToRequest given no request body should translate parameters 1`] = `
Object {
  "body": Object {
    "contents": Array [],
  },
  "cookie": Array [],
  "headers": Array [
    Object {
      "content": Object {
        "content-c": Object {
          "schema": Object {},
        },
      },
      "description": "descr",
      "examples": Array [],
      "name": "param-name-3",
    },
  ],
  "path": Array [],
  "query": Array [
    Object {
      "content": Object {
        "content-a": Object {
          "schema": Object {},
        },
      },
      "deprecated": true,
      "description": "descr",
      "examples": Array [],
      "name": "param-name-1",
    },
    Object {
      "content": Object {
        "content-b": Object {
          "schema": Object {},
        },
      },
      "description": "descr",
      "examples": Array [],
      "name": "param-name-2",
    },
  ],
}
`;
